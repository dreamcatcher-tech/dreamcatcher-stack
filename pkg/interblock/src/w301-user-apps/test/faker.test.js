import { apps, Interpulse } from '../../index.mjs'
import { dirname } from 'path'
import { fileURLToPath } from 'url'
import Debug from 'debug'
import fs from 'fs'
const debug = Debug('tests')

describe('faker', () => {
  describe('routing', () => {
    test('single', async () => {
      const { routing } = await import('../src/crm/faker/index.js')
      expect(routing.length).toEqual(40)

      const engine = await Interpulse.createCI({
        overloads: { '/crm': apps.crm.covenant },
      })
      await engine.add('routing', '/crm/routing')
      debug('routing', routing[0])
      await engine.execute('routing/add', routing[0])
      const sector = await engine.latest('routing/0')
      const state = sector.getState().toJS()
      expect(state.formData).toEqual(routing[0].formData)
      expect(state).toMatchSnapshot()
    })
    test('batch', async () => {
      const { routing } = await import('../src/crm/faker/index.js')
      expect(routing.length).toEqual(40)
      const engine = await Interpulse.createCI({
        overloads: { '/crm': apps.crm.covenant },
      })
      await engine.add('routing', '/crm/routing')
      await engine.execute('routing/batch', { batch: routing })
      let index = 0
      for (const { formData } of routing) {
        const sector = await engine.latest(`routing/${index}`)
        const state = sector.getState().toJS()
        expect(state.formData).toEqual(formData)
        index++
      }
    })
  })
  describe('customers', () => {
    test('single', async () => {
      const { customers } = await import('../src/crm/faker/index.js')
      const single = customers.generateSingle()
      debug('single', single)

      const engine = await Interpulse.createCI({
        overloads: { '/crm': apps.crm.covenant },
      })
      await engine.add('customers', '/crm/customers')
      debug('customers', single)
      await engine.execute('customers/add', single)
      const sector = await engine.current('customers/' + single.formData.custNo)
      const state = sector.getState().toJS()
      expect(state.formData).toEqual(single.formData)
      expect(state).toMatchSnapshot()
    })
    test('batch', async () => {
      const { customers } = await import('../src/crm/faker/index.js')
      const batchSize = 11
      const batch = customers.generateBatch(batchSize)
      expect(batch.length).toEqual(batchSize)
      const engine = await Interpulse.createCI({
        overloads: { '/crm': apps.crm.covenant },
      })
      await engine.add('customers', '/crm/customers')
      await engine.execute('customers/batch', { batch })
      for (const { formData } of batch) {
        const sector = await engine.current(`customers/${formData.custNo}`)
        const state = sector.getState().toJS()
        expect(state.formData).toEqual(formData)
      }
    })
  })
})

describe.skip('generators', () => {
  // TODO generate using the installer, then add new customers
  // using a harness

  const write = (complex, filename) => {
    delete complex.tree
    const path = `${__dirname}/${filename}.js`
    fs.writeFileSync(path, 'export default ' + JSON.stringify(complex, null, 2))
  }
  const __dirname = dirname(fileURLToPath(import.meta.url))
  test('small', async () => {
    const { default: faker } = await import('../src/crm/faker')
    const complex = faker(100)
    const customers = complex.child('customers')
    expect(customers.network.length).toEqual(100)
    write(complex, 'small')
  })
  test('medium', async () => {
    const { default: faker } = await import('../src/crm/faker')
    const complex = faker(1000)
    const customers = complex.child('customers')
    expect(customers.network.length).toEqual(1000)
    write(complex, 'medium')
  })
  test('large', async () => {
    const { default: faker } = await import('../src/crm/faker')
    const complex = faker(20000)
    const customers = complex.child('customers')
    expect(customers.network.length).toEqual(20000)
    write(complex, 'large')
  })
})
