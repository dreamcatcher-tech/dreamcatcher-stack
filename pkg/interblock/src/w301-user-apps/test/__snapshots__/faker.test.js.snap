// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`faker customers single 1`] = `
{
  "formData": {
    "custNo": 777,
    "importedHash": "171e08cfb3ddfabbabbe2fec3beab29a54baa4c2",
    "mobile": "+61 4 7854 5398",
    "name": "Raquel Ryan",
    "serviceAddress": "162 Cole Trail",
    "serviceGps": {
      "latitude": -37.918212,
      "longitude": 175.253737,
    },
  },
  "schema": "..",
}
`;

exports[`faker routing single 1`] = `
{
  "formData": {
    "color": "cyan",
    "frequencyInDays": 7,
    "frequencyOffset": 0,
    "geometry": {
      "geometry": {
        "coordinates": [
          [
            [
              175.25263667062976,
              -37.7440816347928,
            ],
            [
              175.25447308995356,
              -37.74285429309597,
            ],
            [
              175.2482092380524,
              -37.74007573162464,
            ],
            [
              175.2468037605286,
              -37.73792916845374,
            ],
            [
              175.2453339099884,
              -37.73694495267064,
            ],
            [
              175.2380704879761,
              -37.73646980933366,
            ],
            [
              175.23473381996155,
              -37.73175214933702,
            ],
            [
              175.22886514663696,
              -37.724955135893936,
            ],
            [
              175.2258932590485,
              -37.7217133941832,
            ],
            [
              175.20791172981265,
              -37.7265335242151,
            ],
            [
              175.21149516105655,
              -37.73596072380298,
            ],
            [
              175.22545337677002,
              -37.74551680793711,
            ],
            [
              175.23420095421903,
              -37.74272279101128,
            ],
            [
              175.2365863331943,
              -37.74666492475831,
            ],
            [
              175.23770570755008,
              -37.74760093824437,
            ],
            [
              175.23989439010623,
              -37.74803359260726,
            ],
            [
              175.2410852909088,
              -37.7475076196756,
            ],
            [
              175.24511933326724,
              -37.746447179011966,
            ],
            [
              175.25025486951563,
              -37.74492435964226,
            ],
            [
              175.25263667062976,
              -37.7440816347928,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": {},
      "type": "Feature",
    },
    "name": "Monday A Run",
  },
  "schema": "..",
}
`;
